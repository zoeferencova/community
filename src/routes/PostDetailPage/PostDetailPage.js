import React, { Component } from "react";
import { withRouter } from "react-router-dom";
import { CHAT_STARTED, MESSAGE_SENT } from "../../message-utils/events";
import { ButtonDark, ButtonLight } from "../../components/Utils/Utils";
import ChatService from "../../services/chat-service";
import CommUnityContext from "../../contexts/context";
import Task from "../../components/Task/Task";
import GoogleMap from "../../components/GoogleMap/GoogleMap";
import { Textarea } from "../../components/Utils/Utils";

import styles from "./PostDetailPage.module.css";


class PostDetailPage extends Component {
    static contextType = CommUnityContext;

    state = { 
        error: null,
        loading: false
    }
    
    findPost() {
        const postId = parseFloat(this.props.match.params.id);
        return this.context.neighborhood_posts.find(post => post.id === postId)
    }

    handleSubmit(e) {
        e.preventDefault();

        this.setState({...this.state, loading: true })

        const post = this.findPost();
        const messageContent = e.target.message.value;

        if (!messageContent.length) {
            this.setState({ error: "Please enter a message", loading: false })
        } else {
            const receiverId = post.user_id;
            const newChat = { user2Id: receiverId, postId: post.id }
            ChatService.postChat(newChat)
                .then(chat => {
                    this.setState({...this.state, loading: false })
                    this.props.history.push("/messages")
                    this.context.addNewChat(chat)
                    this.context.socket.emit(CHAT_STARTED, { receiverId, chat })
                    const newMessage = { message_content: messageContent, chat_id: chat.id  }
                    ChatService.postMessage(newMessage)
                        .then(message => {
                            this.context.addNewMessage(message, message.chat_id)
                            this.context.socket.emit(MESSAGE_SENT, { sender: this.context.user, receiverId, message })
                        })
                })
        }
    }

    goToMessages = chat => {
        this.context.updateActiveChat(chat.id)
        this.props.history.push("/messages")
    }

    render() {
        const post = this.findPost();

        return (   
            <main className={styles.main}>
                {post && <>
                    <div className={styles.map}>
                        <GoogleMap className={styles.gmap} location={post.location} radius={post.radius} displayMarker={false} />
                    </div>
                    <div className={styles.postHeader}>
                        <h3 className={styles.postTitle}>Respond to {post.first_name}'s {post.post_type}</h3>
                        {post.post_type === "request" && <span className={styles[post.urgency]}><i className="fas fa-circle"></i> {post.urgency} urgency</span>}
                    </div>
                    <ul className={styles.tasks}>
                        {post.categories.map(task => <Task key={task} task={task} />)}
                    </ul>
                    {post.description && <p className={styles.description}>{post.description}</p>}
                    <div className={styles.messageSection}>
                        {!this.context.chats.find(chat => chat.user1.id === post.user_id || chat.user2.id === post.user_id) ?
                            <form className={styles.form} onSubmit={e => this.handleSubmit(e)}>
                                <label htmlFor="message">Write a message</label>
                                <Textarea id="message" className={`${styles.textarea} ${this.state.error && styles.errorCell}`} placeholder={`Hi ${post.first_name}...`}></Textarea>
                                {this.state.error && <div className={styles.error}>{this.state.error}</div>}
                                <div className={styles.buttonSection}>
                                    <ButtonLight type="button" onClick={() => this.props.history.goBack()}>Back</ButtonLight>
                                    <ButtonDark type="submit" className={styles.submitButton} loading={this.state.loading.toString()}>Send Message</ButtonDark>
                                </div>
                            </form>
                            :
                            <>
                                <p className={styles.chatMessage}>You have a chat with {post.first_name}</p> 
                                <div className={styles.buttonSection}>
                                        <ButtonLight type="button" onClick={() => this.props.history.goBack()}>Back</ButtonLight>
                                        <ButtonDark type="button" onClick={() => this.goToMessages(this.context.chats.find(chat => chat.user1.id === post.user_id || chat.user2.id === post.user_id))}>Go to chat</ButtonDark>
                                </div>  
                            </>
                        }
                    </div>
                    
                </>}
            </main>
        )
    }
}

export default withRouter(PostDetailPage);