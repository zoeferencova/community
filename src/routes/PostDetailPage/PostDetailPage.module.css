.postHeader {
    display: flex;
    justify-content: space-between;
    align-items: center;
    margin-top: 30px;
}

.details {
    width: 85%;
}

.postHeader h3 {
    font-size: 18px;
    font-weight: 500;
}

.postHeader span {
    display: inline-block;
    font-size: 13px;
}

.postHeader svg {
    opacity: 0.6;
    margin-right: 8px;
}

.high svg {
    color: #ad3434;
}

.medium svg {
    color: #edcb64;
}

.low svg {
    color: #3d9145;
}

.urgency {
    text-transform: capitalize;
}

.tasks {
    display: flex;
    list-style: none;
    padding-left: 0px;
    flex-wrap: wrap;
    margin-top: 5px;
}

.map {
    height: 250px;
    width: 100% !important;
    background-color: #ddd;
}

.textarea {
    height: 100px;
    margin: 15px 0px;
}

.form {
    display: flex;
    flex-direction: column;
}

.form label {
    align-self: flex-start;
    font-size: 14px;
}

.messageSection {
    align-self: center;
    margin: 20px 0px 0px 0px;
}

.buttonSection {
    display: flex;
    justify-content: space-between;
    margin-top: 10px;
}

.submitButton {
    min-width: 120px;
}

.postTitle {
    font-size: 18px;
    text-align: left;
}

.description {
    font-size: 15px;
    text-align: left;
    margin: 10px 0px 25px 0px;
    line-height: 1.26em;
}

.chatMessage {
    font-size: 15px;
    padding: 0px 0px 5px 5px;
    border-radius: 5px;
    margin: -5px 0 30px 0;
    color: #8492a4
}

.chatMessage svg {
    margin-right: 5px;
    font-size: 16px;
}

.error {
    text-align: left;
    font-size: 12px;
    color: #a34343;
    margin: -5px 0px 15px 0px;
}

.errorCell {
    border: 1px solid #a34343;
    border-radius: 3px;
}

@media only screen and (min-width: 830px) {
    .description {
        margin: 15px 0 35px 0;
    }

    .textarea {
        height: 45px !important;
    }
}