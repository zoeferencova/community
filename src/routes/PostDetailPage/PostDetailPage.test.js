import React from "react";
import { render } from "enzyme";
import PostDetailPage from "./PostDetailPage";
import { MemoryRouter, Route } from "react-router-dom";
import CommUnityContext from "../../contexts/context";

import mockContext from "../../contexts/mock-context";

describe("PostDetailPage", () => {
    const mockProps = {
        chats: [
            {
                ...mockContext.chats[0], 
                user1: {first_name: "Zoe", id: 1},
                user2: {first_name: "James", id: 2}
            }
        ],
        activeChat: mockContext.activeChat,
        setActiveChat: jest.fn()
    }

    it("renders PostDetailPage component given mock context data", () => { 
        const wrapper = render(
            <CommUnityContext.Provider value={mockContext}>
                <MemoryRouter initialEntries={[`/post/4`]}>
                    <Route path="/post/:id">
                        <PostDetailPage {...mockProps} />
                    </Route>
                </MemoryRouter>
            </CommUnityContext.Provider>,
        );
        
        expect(wrapper).toMatchSnapshot();
    });
});