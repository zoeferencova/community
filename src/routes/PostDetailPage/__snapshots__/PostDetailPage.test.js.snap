// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostDetailPage renders PostDetailPage component given mock context data 1`] = `
<main
  class="main"
>
  <div
    class="map"
  >
    <div
      id="google-map"
      style="height:100%;width:100%"
    />
  </div>
  <div
    class="postHeader"
  >
    <h3
      class="postTitle"
    >
      Respond to James's offer
    </h3>
  </div>
  <ul
    class="tasks"
  >
    <li
      class="task supplies"
    >
      Picking up supplies
    </li>
  </ul>
  <div
    class="messageSection"
  >
    <p
      class="chatMessage"
    >
      You have a chat with James
    </p>
    <div
      class="buttonSection"
    >
      <button
        class="button "
        type="button"
      >
        Back
      </button>
      <button
        class="buttonDark "
        type="button"
      >
        Go to chat
      </button>
    </div>
  </div>
</main>
`;
